okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java

```diff
@@ -82,26 +107,34 @@ public CallServerInterceptor(boolean forWebSocket) {
         .receivedResponseAtMillis(System.currentTimeMillis())
         .build();
 
+    if ("close".equalsIgnoreCase(response.request().header("Connection"))
+        || "close".equalsIgnoreCase(response.header("Connection"))) {
+      streamAllocation.noNewStreams();
+    }
+
     int code = response.code();
     if (forWebSocket && code == 101) {
       // Connection is upgrading, but we need to ensure interceptors see a non-null response body.
       response = response.newBuilder()
           .body(Util.EMPTY_RESPONSE)
           .build();
     } else {
-      response = response.newBuilder()
-          .body(httpCodec.openResponseBody(response))
-          .build();
-    }
+      realChain.eventListener().responseBodyStart(realChain.call());
+      try {
+        response = response.newBuilder()
+            .body(httpCodec.openResponseBody(response))
```

Comment:
I think we wanna move responseBodyStart() and responseBodyEnd() calls into the codec. Main reason is that if there’s exceptions reading the response body these will be the best places to signal the failure to responseBodyEnd.